lib/upsert/merge_function/PG_Connection.rb

```diff
@@ -6,7 +6,70 @@ class MergeFunction
     class PG_Connection < MergeFunction
       include Postgresql
 
+      attr_reader :quoted_setter_names
+      attr_reader :quoted_selector_names
+
+      def self.use_pg_upsert_cache
+        @use_pg_upsert_cache ||= {}
+      end
+
+      def initialize(controller, *args)
+        super
+        @quoted_setter_names = setter_keys.map { |k| connection.quote_ident k }
+        @quoted_selector_names = selector_keys.map { |k| connection.quote_ident k }
+      end
+
       def execute(row)
+        use_pg_upsert? ? pg_upsert(row) : pg_function(row)
+      end
+
+      def use_pg_upsert?
+        return if @use_native_upsert == false
+        # We need to memoize this way because we're storing a boolean
+        return self.class.use_pg_upsert_cache[__id__] if self.class.use_pg_upsert_cache.key?(__id__)
+
+        self.class.use_pg_upsert_cache[__id__] = begin
+          version = controller.connection.execute("SHOW server_version").getvalue(0, 0)
+          version_number = version[0..2].split('.').join('').to_i
```

Comment:
you should return early if the version number is too low - don't bother checking for the presence of indexes